day12/main.go

```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	aoc "github.com/golang-insiders/advent-of-code-2023/library"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Println("input file path not provided")
+		return
+	}
+	input := aoc.ReadFileLineByLine(os.Args[1])
+	fmt.Println("answer for part 1: ", ans(input, 1))
+	fmt.Println("answer for part 2: ", ans(input, 5))
+}
+
+func ans(input []string, times int) int {
+	sum := 0
+	cache := map[string]int{}
+	count := 0
+	for _, line := range input {
+		count, cache = processLine(line, cache, times)
+		sum += count
+	}
+	return sum
+}
+
+func processLine(input string, cache map[string]int, times int) (int, map[string]int) {
+	stringRep, numRep := fetchInputs(input, times)
+	return getCount(stringRep, numRep, cache)
+}
+
+func fetchInputs(input string, times int) ([]string, []int) {
+	stringRep := input[:strings.Index(input, " ")]
+	numRep := aoc.FetchSliceOfIntsInString(input[strings.Index(input, " "):])
+
+	numRep2 := []int{}
+	var build strings.Builder
+	for i := 0; i < times; i++ {
+		build.WriteString(stringRep)
+		if i != (times - 1) {
+			build.WriteString("?")
+		}
+		numRep2 = append(numRep2, numRep...)
+	}
+	stringRep = build.String()
+	return strings.Split(stringRep, ""), numRep2
+}
+
+// top level approach used:
+// - chose direction to go based on first character of the input string:
+//   - '.' -> ignore
+//   - '#' -> look ahead if consequetive '#'s satisfy the first group.
+//   - '?' -> 2 recursive call by replacing once with '.' and then '#'.
+//
+// - to reduce the number of recursive calls, we also have a cache to store all evaluated results.
+func getCount(line []string, group []int, cache map[string]int) (int, map[string]int) {
```

Comment:
You don't have to return the cache struct. It's a map, you can just update the values and be it; it'll work and makes the code simpler.

Reply:
Ah ofcourse, this was silly of me. Thanks for pointing it out. Addressed!